frontend/src/frontend_benchmark/source_generator.py

```diff
@@ -0,0 +1,185 @@
+                 main_file: str = 'main.c', benchmark_name: str = 'benchmark') -> None:
+        self.output_dir: str = output_directory
+        self.callgraph: Callgraph = callgraph
+        self.header_file: str = header_file
+        self.main_file: str = main_file
+        self.benchmark_name: str = benchmark_name
+
+    def write_files(self, splits: int = None) -> None:
+        self.write_main()
+        self.write_headers()
+        function_files = self.write_functions(splits)
+        self.write_makefile(function_files)
+
+    def write_main(self) -> None:
+        template = self._build_main_template()
+        main_path = self.append_path_to_output_dir(self.main_file)
+        with open(main_path, 'w') as f:
+            f.write(template)
+
+    def _build_main_template(self) -> str:
+        header = self.get_header_import_string()
+        vars_def = self.callgraph.format_vars_definition()
+        function_call = self.callgraph.function_call_signature_for(self.callgraph.entry_point)
+        variable = 'loops'
+        arg_template = self._build_arg_template(variable)
+        template = (
+            f'#include <unistd.h>\n'
+            f'#include <stdio.h>\n'
+            f'#include <stdlib.h>\n'
+            f'{header}\n\n'
+            f'{vars_def}\n'
+            'int main(int argc, char **argv) {\n'
+            f'unsigned long {variable} = 1;\n'
+            f'{arg_template}\n'
+            f'for (int i = 0; i < {variable}; i++)'
+            ' {\n'
+            f'{function_call}();\n'
+            '}\n'
+            '}\n'
+        )
+        return template
+
+    def get_header_import_string(self) -> str:
+        return f'#include "{self.header_file}"'
+
+    def _build_arg_template(self, variable) -> str:
+        template = (
+            'int c;\n'
+            'while ((c = getopt(argc, argv, "l:")) != -1) {\n'
+            'switch (c) {\n'
+            "case 'l':\n"
+            f'{variable} = strtoul(optarg, NULL, 0);\n'
+            '}\n'
+            '}'
+        )
+        return template
+
+    def append_path_to_output_dir(self, path) -> str:
+        return os.path.join(self.output_dir, path)
+
+    def write_headers(self) -> None:
+        header_path = self.append_path_to_output_dir(self.header_file)
+        with open(header_path, 'w') as f:
+            f.write(self.callgraph.format_vars_declaration())
+            f.write(self.callgraph.format_headers())
+
+    def write_functions(self, splits: int = None) -> Collection[str]:
+        grouper = FileFunctionMapper(self.callgraph)
+        file_to_function_name_mapping = grouper.group_functions(splits)
+        for function_file, functions in file_to_function_name_mapping.items():
+            self.write_header_import(function_file)
+            for function_name in functions:
+                self.write_function(function_name, function_file)
+        return file_to_function_name_mapping.keys()
+
+    def write_header_import(self, path) -> None:
```

Comment:
The pattern here is to just open an existing file and either write or append to it, which is fine, but the required ordering is implicit, so it's easy for future maintainers to get wrong. I'd suggest two things:

1. name functions with "write_" or "append_" based on their behavior.
2. enforce the ordering with a pattern like "close_headers", which flips a flag that will raise an error if the user attempts to call it again afterwards.
3. some documentation.

Reply:
Good point. I changed the function names to be more descriptive, and I also added a check in write headers to ensure the file does not already exist.